build.py

```diff
@@ -0,0 +1,300 @@
+
+    def __exit__(self, exception, value, tb):
+        self.busy = False
+        time.sleep(self.delay)
+        if exception is not None:
+            return False
+
+
+def run_subprocess_with_spinner(name: str, proc: subprocess.Popen):
+    """
+    Opens a subprocess and redirect stdout and stderr to Python.\n
+    Shows a spinning Cursor while the command runs.\n
+    Exits with returncode of subprocess if not equals 0.\n
+    """
+    try:
+        p = None
+        # Register handler to pass keyboard interrupt to the subprocess
+
+        def handler(sig, frame):
+            print(f"===================== {
+                  name} ABORTED BY USER =========================")
+            if p:
+                p.send_signal(signal.SIGINT)
+            else:
+                raise KeyboardInterrupt.add_note()
+        signal.signal(signal.SIGINT, handler)
+        with Spinner():
+            print(f"{name.rstrip('Ee')}ing...")
+            with proc as p:
+                for line in iter(p.stdout.readline, b''):
+                    # Print the line to the console
+                    sys.stdout.buffer.write(line)
+                for line in iter(p.stderr.readline, b''):
+                    # Print the error output to the console
+                    sys.stderr.buffer.write(line)
+                returncode = p.wait()
+                if returncode:
+                    print(f"===================== {
+                        name} FAILED =========================")
+                    sys.exit(returncode)
+                else:
+                    print(f"===================== {
+                          name} SUCCESSFUL =========================")
+    finally:
+        signal.signal(signal.SIGINT, signal.SIG_DFL)
+
+
+def get_active_branch_name():
+    head_dir = pathlib.Path(DIR_PATH + "/.git/HEAD")
+    with head_dir.open("r") as f:
+        content = f.read().splitlines()
+
+    for line in content:
+        if line[0:4] == "ref:":
+            return line.partition("refs/heads/")[2]
+
+
+class Build:
+    def __init__(self, openstack: bool, template: str, conda_env: pathlib.Path, packer_path: pathlib.Path, provisioning: [str], comment: str, pvt_key: pathlib.Path, ansible_args: str):
+        self.openstack = openstack
+        self.template = template
+        self.os = "-".join(template.split("-", 2)[:2])
+        self.comment = comment
+        self.pvt_key = pvt_key
+        self.conda_env = conda_env
+        self.provisioning = provisioning
+        self.ansible_args = ansible_args
+        self.packer_path = str(conda_env) + \
+            "/bin/packer" if conda_env != None else str(
+                packer_path) | shutil.which("qemu-img")
+        self.qemu_path = str(conda_env) + \
+            "/bin/qemu-img" if conda_env != None else shutil.which("qemu-img")
+        self.openstack_path = str(conda_env) + \
+            "/bin/openstack" if conda_env != None else shutil.which(
+                "openstack")
```

Comment:
I disagree with @bgruening's [comment](https://github.com/usegalaxy-eu/vgcn/pull/80/files#r1428926940), I would argue this four-liner is readable enough.

```suggestion
        self.binaries = {
            conda_env / command if conda_env else shutil.which(name)
            for command in ("packer", "qemu-img", "openstack")
        }
```

Reply:
Update: I like the current solution too.